### DISCUSSION TITLE
Need some pointers on how to diagnose OOM exception

### DISCUSSION THREAD

**Discussion by En3Tho** [tailwindlabs/tailwindcss#9735] [Help]
**What version of Tailwind CSS are you using?**

v3.2.1

**What build tool (or framework if it abstracts the build tool) are you using?**

"devDependencies": {
    "cross-env": "^7.0.3",
    "autoprefixer": "^10.4.7",
    "postcss": "^8.4.14",
    "postcss-cli": "^10.0.0",
    "tailwindcss": "^3.2.1",
    "postcss-nested": "^6.0.0",
    "postcss-custom-media": "^8.0.2",
    "postcss-import": "^15.0.0"
}

**What version of Node.js are you using?**

v18.2.0

**What operating system are you using?**

Windows

**Describe your issue**

Describe the problem you're seeing, any important steps to reproduce and what behavior you expect instead.

I've started playing with Tailwind and Blazor recently and I've stumbled upon a bug: 
```npx tailwindcss -i ... --watch```

always ends up with

```
<--- Last few GCs --->

[54384:000002442C6BFB10]  3614162 ms: Mark-sweep (reduce) 4071.2 (4143.1) -> 4070.8 (4143.6) MB, 3521.8 / 4.2 ms  (average mu = 0.336, current mu = 0.003) allocation failure scavenge might not succeed
[54384:000002442C6BFB10]  3617660 ms: Mark-sweep (reduce) 4072.0 (4143.6) -> 4071.3 (4144.1) MB, 3496.6 / 4.2 ms  (average mu = 0.197, current mu = 0.000) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF663DC30AF v8::internal::CodeObjectRegistry::~CodeObjectRegistry+112511
 2: 00007FF663D52216 DSA_meth_get_flags+65542
 3: 00007FF663D530CD node::OnFatalError+301
 4: 00007FF664684B6E v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF66466F09D v8::SharedArrayBuffer::Externalize+781
 6: 00007FF66451268C v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF66450F7C4 v8::internal::Heap::CollectGarbage+4244
 8: 00007FF66450D140 v8::internal::Heap::AllocateExternalBackingStore+2000
 9: 00007FF664531A76 v8::internal::Factory::NewFillerObject+214
10: 00007FF664264CB5 v8::internal::DateCache::Weekday+1797
11: 00007FF664712541 v8::internal::SetupIsolateDelegate::SetupHeap+494417
12: 00007FF664713449 v8::internal::SetupIsolateDelegate::SetupHeap+498265
```

When I start tailwind jit it slowly goes from like 25ms incremental rebuild to 1sec+ and then fails with this exception. Stacktrace can be little bit different each time. But it always reaches 4gb and fails. My project is quite small - it's a playground basically, not a full fledged app with hundreds of components.

4bg seems like an artificial limit to me but I'm not sure it even should reach this high. I have enough memory on my machine (64 gb) so I can assure I have quite a lot memory left when fail occurs.

I'd like to get some pointers on how to debug this kind of thing? I'd really like to get this fixed.

**Comment by adamwathan** [comment]
Hey! Need a reproduction to be able to help.

**Reply by En3Tho** [reply]
Thank you for replying. I will try to clean up a little bit and come with an instructions how to run my app.

I've noticed I had .css files included in my file match pattern for some reason. Removing that and leaving only .html, cshtml,. razor helped a little bit but it still failed with oom. Funnily, one of rebuilds took outstanding 48 seconds. But general behavior is like start with 25ms, gradually get to about 300ms and fail with oom.

Are you familiar with .Net/Blazor by the way?

**Comment by LakeQueen** [comment]
Hi Adam, just letting you know I have a similar problem on a Next 13 app. No glob patterns in config, but the JIT compiler slowly runs out of heap memory over the course of several hours in --watch mode. This has happened 3 times so far on 2 separate projects, which makes me believe it is a general issue. However, I don't know how to reproduce it besides repeatedly saving files until it crashes eventually.

Perhaps if the JIT has a --debug option I can be more helpful?

**Comment by En3Tho** [comment]
I don't think I have mental capacity to profile JS after working with .Net tools but I can say for sure that there are LOTS of duplicate strings like these (note 212k of those):

![image](https://user-images.githubusercontent.com/37334640/204565936-3cab27e0-21f9-44c5-90ca-86d2b882d9ca.png)

Amount is getting bigger every time.
So maybe the issue is somewhere around the cache. I've also noticed I've got node_modules folder included in glob:

```
 content: [
        './**/*.{razor,html,cshtml}',
        './../FSharpComponents/**/*.fs',
    ],
```

But I doubt this should lead to OOM. Just slow things down.

GC goes crazy after a while too:
![image](https://user-images.githubusercontent.com/37334640/204572275-8adbe261-5e44-473b-ab01-a4f6f5bddbc1.png)

**Comment by RyanTT** [comment]
I am experiencing the same issue. I am running the Tailwind JIT together with .NET (dotnet watch). The builds slowly get slower and slower until it eventually crashes. I have also noticed that, when running the Tailwind JIT with --watch, the dotnet tool will eventually be able to apply Hot-Reload changes (this only happens when running with Tailwind JIT).

**Reply by 0ptim** [reply]
Hey @RyanTT, did you find a solution to your problem or is this still pending? I'm currently facing similar issues.

**Reply by En3Tho** [reply]
The most straightforward way is to use a script that just restarts the process and spare yourself time wasted on js tooling. I don't know if this issue was fixed but seeing you asking for the solution I guess it wasn't.
